lib/pac55xx/ccs.c

```diff
@@ -0,0 +1,264 @@
+
+static uint32_t ccs_extclk_frequency = CCS_ROSC_FREQ;
+static uint32_t ccs_frclk_frequency = CCS_ROSC_FREQ;
+static uint32_t ccs_sclk_frequency = CCS_ROSC_FREQ;
+static uint32_t ccs_pll_clk_frequency = 0;
+static uint32_t ccs_hclk_frequency = CCS_ROSC_FREQ;
+static uint32_t ccs_aclk_frequency = CCS_ROSC_FREQ;
+static uint32_t ccs_pclk_frequency = CCS_ROSC_FREQ;
+
+void ccs_frclkmux_select(uint32_t sel) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_FRCLKMUXSEL(CCS_CTL_FRCLKMUXSEL_MASK)) | CCS_CTL_FRCLKMUXSEL(sel);
+}
+void ccs_rosc_enable(void) {
+	CCSCTL |= CCS_CTL_ROSCEN;
+}
+void ccs_rosc_disable(void) {
+	CCSCTL &= ~CCS_CTL_ROSCEN;
+}
+void ccs_sclkmux_select_frclk(void) {
+	CCSCTL &= ~CCS_CTL_SCLKMUXSEL;
+}
+void ccs_sclkmux_select_pllclk(void) {
+	CCSCTL |= CCS_CTL_SCLKMUXSEL;
+}
+void ccs_clkfail_enable(void) {
+	CCSCTL |= CCS_CTL_CLKFAILEN;
+}
+void ccs_clkfail_disable(void) {
+	CCSCTL &= ~CCS_CTL_CLKFAILEN;
+}
+void ccs_clkfailmux_select_frclk(void) {
+	CCSCTL &= ~CCS_CTL_CLKFAILMUXSEL;
+}
+void ccs_clkfailmux_select_pllclk(void) {
+	CCSCTL |= CCS_CTL_CLKFAILMUXSEL;
+}
+void ccs_ldo_enable(void) {
+	CCSCTL |= CCS_CTL_LDOEN;
+}
+void ccs_ldo_disable(void) {
+	CCSCTL &= ~CCS_CTL_LDOEN;
+}
+void ccs_pclk_enable(void) {
+	CCSCTL |= CCS_CTL_PCLKEN;
+}
+void ccs_pclk_disable(void) {
+	CCSCTL &= ~CCS_CTL_PCLKEN;
+}
+void ccs_aclk_enable(void) {
+	CCSCTL |= CCS_CTL_ACLKEN;
+}
+void ccs_aclk_disable(void) {
+	CCSCTL &= ~CCS_CTL_ACLKEN;
+}
+void ccs_adcclk_enable(void) {
+	CCSCTL |= CCS_CTL_ADCCLKEN;
+}
+void ccs_adcclk_disable(void) {
+	CCSCTL &= ~CCS_CTL_ADCCLKEN;
+}
+void ccs_stclk_sleep_enable(void) {
+	CCSCTL |= CCS_CTL_STCLKSLPEN;
+}
+void ccs_stclk_sleep_disable(void) {
+	CCSCTL &= ~CCS_CTL_STCLKSLPEN;
+}
+void ccs_set_pclkdiv(uint32_t div) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_PCLKDIV(CCS_CTL_PCLKDIV_MASK)) | CCS_CTL_PCLKDIV(div);
+}
+void ccs_set_aclkdiv(uint32_t div) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_ACLKDIV(CCS_CTL_ACLKDIV_MASK)) | CCS_CTL_ACLKDIV(div);
+}
+void ccs_set_hclkdiv(uint32_t div) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_HCLKDIV(CCS_CTL_HCLKDIV_MASK)) | CCS_CTL_HCLKDIV(div);
+}
+void ccs_pll_enable(void) {
+	CCSPLLCTL |= CCS_PLLCTL_PLLEN;
+}
+void ccs_pll_disable(void) {
+	CCSPLLCTL &= ~CCS_PLLCTL_PLLEN;
+}
+bool ccs_pll_locked(void) {
+	return (CCSPLLCTL & CCS_PLLCTL_PLLLOCK) == CCS_PLLCTL_PLLLOCK;
+}
+void ccs_pll_bypass_enable(void) {
+	CCSPLLCTL |= CCS_PLLCTL_PLLBP;
+}
+void ccs_pll_bypass_disable(void) {
+	CCSPLLCTL &= ~CCS_PLLCTL_PLLBP;
+}
+void ccs_pll_set_outdiv(uint32_t div) {
+	CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLOUTDIV(CCS_PLLCTL_PLLOUTDIV_MASK)) | CCS_PLLCTL_PLLOUTDIV(div);
+}
+void ccs_pll_set_indiv(uint32_t div) {
+	if (div <= 15 && div >= 1) {
+		CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLINDIV(CCS_PLLCTL_PLLINDIV_MASK)) | CCS_PLLCTL_PLLINDIV(div);
+	}
+}
+void ccs_pll_set_fbdiv(uint32_t div) {
+	if (div <= 16383 && div >= 4) {
+		CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLFBDIV(CCS_PLLCTL_PLLFBDIV_MASK)) | CCS_PLLCTL_PLLFBDIV(div);
+	}
```

Comment:
if you're going to bother error checking, it would seem reasonable to provide a return code, but that's very rarely done anywhere else either so *shrugs*

Reply:
yeah, somewhat unrelated: found that there isn't a debug mechanism to unbrick a chip if you get the clock config wrong.  so we put in some software hooks outside of libopencm3 to avoid bricking.  regardless, I just wanted to prevent writing an invalid value to a register and let the user debug why it isn't working. they'd figure it out quickly with GDB

Reply:
I put in `cm3_assert_not_reached();` if the parameters are bad - to be similar to the rest of this source file